File: LeetCode-POTD-Solutions/Reorder-Routes-to-Make-All-Paths-Lead-to-the-City-Zero.cpp
1class Solution {
2public:
3    int dfs(int node, int parent, vector<vector<pair<int,int>>>& graph) {
4        int changes = 0;
5
6        for (auto& [next, cost] : graph[node]) {
7            if (next == parent) continue;
8            changes += cost;
9            changes += dfs(next, node, graph);
10        }
11
12        return changes;
13    }
14
15    int minReorder(int n, vector<vector<int>>& connections) {
16        vector<vector<pair<int,int>>> graph(n);
17
18        for (auto& c : connections) {
19            int a = c[0], b = c[1];
20            graph[a].push_back({b, 1}); // original direction, needs change
21            graph[b].push_back({a, 0}); // reverse direction, already fine
22        }
23
24        return dfs(0, -1, graph);
25    }
26};
27

File: LeetCode-POTD-Solutions/Reorder-Routes-to-Make-All-Paths-Lead-to-the-City-Zero.py
class Solution(object):
    def minReorder(self, n, connections):
        from collections import defaultdict, deque
        
        graph = defaultdict(list)
        
        # Build graph: store directed edges and undirected edges
        for a, b in connections:
            graph[a].append((b, 1))  # edge originally a -> b
            graph[b].append((a, 0))  # reverse edge b -> a (no need to reorder)
        
        q = deque([0])
        visited = set([0])
        changes = 0
        
        while q:
            node = q.popleft()
            for nei, cost in graph[node]:
                if nei not in visited:
                    visited.add(nei)
                    changes += cost
                    q.append(nei)
        
        return changes
